From: Joerg Roedel <joerg.roedel@amd.com>
Subject: [PATCH 2/5] KVM: X86: Let kvm-clock report the right tsc frequency
Reference: FATE#309762
Git-commit: 1e993611d0dc879fde25515dc9867d1cfd4c5137
Patch-mainline: v3.0-rc1

This patch changes the kvm_guest_time_update function to use
TSC frequency the guest actually has for updating its clock.

Signed-off-by: Joerg Roedel <joerg.roedel@amd.com>
Acked-by: Bruce Rogers <brogers@suse.com>
---
 arch/x86/include/asm/kvm_host.h |    2 ++
 arch/x86/kvm/x86.c              |   10 +++++++++-
 2 files changed, 11 insertions(+), 1 deletions(-)

Index: linux-2.6.32-SLE11-SP2/arch/x86/include/asm/kvm_host.h
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/include/asm/kvm_host.h
+++ linux-2.6.32-SLE11-SP2/arch/x86/include/asm/kvm_host.h
@@ -347,6 +347,8 @@ struct kvm_vcpu_arch {
 	u64 last_guest_tsc;
 	u64 last_kernel_ns;
 
+	u32 virtual_tsc_khz;
+
 	bool nmi_pending;
 	bool nmi_injected;
 
Index: linux-2.6.32-SLE11-SP2/arch/x86/kvm/x86.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/kvm/x86.c
+++ linux-2.6.32-SLE11-SP2/arch/x86/kvm/x86.c
@@ -696,6 +696,14 @@ static void kvm_set_time_scale(uint32_t
 
 static DEFINE_PER_CPU(unsigned long, cpu_tsc_khz);
 
+static u64 vcpu_tsc_khz(struct kvm_vcpu *vcpu)
+{
+	if (vcpu->arch.virtual_tsc_khz)
+		return vcpu->arch.virtual_tsc_khz;
+	else
+		return __get_cpu_var(cpu_tsc_khz);
+}
+
 static void kvm_write_guest_time(struct kvm_vcpu *v)
 {
 	struct timespec ts;
@@ -709,7 +717,7 @@ static void kvm_write_guest_time(struct
 	if ((!vcpu->time_page))
 		return;
 
-	this_tsc_khz = get_cpu_var(cpu_tsc_khz);
+	this_tsc_khz = vcpu_tsc_khz(v);
 	if (unlikely(vcpu->hv_clock_tsc_khz != this_tsc_khz)) {
 		kvm_set_time_scale(this_tsc_khz, &vcpu->hv_clock);
 		vcpu->hv_clock_tsc_khz = this_tsc_khz;
